datasource/ethereum/src/block_ingestor.rs

```diff
@@ -153,28 +153,66 @@ where
 
     fn get_latest_block<'a>(
         &'a self,
-    ) -> impl Future<Item = Block<Transaction>, Error = Error> + 'a {
+    ) -> impl Future<Item = Block<Transaction>, Error = BlockIngestorError> + 'a {
         let web3 = Web3::new(self.web3_transport.clone());
         web3.eth()
             .block_with_txs(BlockNumber::Latest.into())
             .map_err(|e| format_err!("could not get latest block from Ethereum: {}", e))
+            .from_err()
             .and_then(|block_opt| {
-                block_opt.ok_or(format_err!("no latest block returned from Ethereum"))
+                block_opt.ok_or(format_err!("no latest block returned from Ethereum").into())
+            })
+    }
+
+    fn load_full_block<'a>(
+        &'a self,
+        block: Block<Transaction>,
+    ) -> impl Future<Item = EthereumBlock, Error = BlockIngestorError> + 'a {
+        let block_hash = block.hash.unwrap();
+
+        // Load transaction receipts
+        let receipt_futures = block
+            .transactions
+            .iter()
+            .map(move |tx| {
+                self.get_transaction_receipt(tx.hash)
+                    .and_then(move |receipt| {
+                        // Check if receipt is for the right block
+                        if receipt.block_hash != block_hash {
+                            // This block is no longer on the main chain according to the Ethereum
+                            // node.  Nothing we can do from here except give up trying to ingest
+                            // this block.
+                            Err(BlockIngestorError::BlockUnavailable(block_hash))
+                        } else {
+                            Ok(receipt)
+                        }
+                    })
+            }).collect::<Vec<_>>();
```

Comment:
is this `collect` necessary?

Reply:
I think so, otherwise they'll load serially instead of parallel

Reply:
I don't follow, `futures_unordered` would execute in parallel.

Reply:
Oh good point, those methods collect implicitly don't they. 
So I just tried to remove the collect and the compiler error now reminds me that I collected here to save having to clone. I need `block` later on to move it into the EthereumBlock, so to avoid collecting I would have to clone a bunch anyway, which seemed like a waste...

Reply:
Weird, well let's not spend too much time on this.